### ISSUE TITLE
[processor/transformprocessor] Allow external functions register to functions in transform processor

### ISSUE THREAD

**Issue by linjunzhe** [open-telemetry/opentelemetry-collector-contrib#16098]
### Component(s)

processor/transform

### Is your feature request related to a problem? Please describe.

Although there are many useful functions in `pkg/ottl/ottlfuncs`, it's not enough for us in some complicated use cases. For example, `/var/log/secure.log` has many different patterns log lines. I want to parse `/var/log/secure.log` with different patterns and extract some key values to `attributes`. 

So I suggest exposing an API, for example, `func Register(name string, function interface{})`, so developers can add their OTTL functions to the transform processor function list. Then use the transform processor to run their business logic. 

It's easier than building a private processor because developers only need to focus on the business logic implementation. 

I can prepare a PR if you think it's worth to do that. 

### Describe the solution you'd like

Expose an API, for example `func Register(name string, function interface{})`, to allow users to register their private functions to the transform processor function list.

**Comment by evan-bradley** [comment]
Hi @linjunzhe, thanks for the detailed request. I think this is an area that could use some attention if transform processor users need highly specific functionality in addition to the functions provided out-of-the-box.

The transform processor already has the basis of functionality you are describing with the `Functions` function in the package for each context. [Here is the function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/internal/logs/functions.go#L22) for the logs context. Custom functions could be used by forking the transform processor and adding them to this map, then building a Collector with the custom processor.

There are two downsides to this approach:
- It requires some knowledge on transform processor implementation details.
- It requires maintaining a fork of the transform processor.

The most seamless solution would likely be to be able to load a binary of compiled functions into the Collector in a way that allows the functions to be injected into the transform processor, but I'm not aware of any mechanisms in the Collector capable of doing this.

Exposing an API that registers functions like you're describing would be an improvement over maintaining a fork since it could be used to wrap the transform processor. That would allow only dealing with external interfaces of the transform processor, with the caveat that the user would still have to maintain a wrapper component. If implemented well, I think this would likely be a workable solution for users. Does wrapping the transform processor agree with your expectations for how you'd like to use a `Register` function?

**Comment by TylerHelmuth** [comment]
@evan-bradley are you imaging the transform processor be extendable?  Like building a customer processor using the transform processor, and all this this custom processor really does is register more functions?

**Comment by evan-bradley** [comment]
Yeah, that's the first approach I can think of that would allow users to easily take advantage of a `Register` function. Since the transform processor is already distributed as a Go package, it would be relatively easy to provide a `NewCustomFactory` or something similar in the transform processor that makes the custom processor a pretty thin layer.

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-contrib-approvers @open-telemetry/collector-contrib-maintainers do we have any components today that can be used as a "base component" for other components?

**Comment by Aneurysm9** [comment]
I think it is reasonable to make the transform processor extensible, but before we go down that path perhaps we should back up and look at the request being made here:

> I want to parse /var/log/secure.log with different patterns and extract some key values to attributes.

This reads to me like a request for log parsing and pattern matching.  Don't we already have receiver components responsible for this type of functionality?  I'm not sure that it belongs in the transform processor, which I would tend to think should be used for processing already parsed/extracted data and not as a general purpose log attribute extraction system.

What guidance do we have at present for component authors to determine whether functionality should be in a receiver, exporter, or the transform processor?  I think we need to have some to avoid the transform processor becoming a general purpose programming language embedded in collector configuration.

**Comment by TylerHelmuth** [comment]
@Aneurysm9 I would say for logs especially that parsing is best done via receivers whenever possible.  For instances when a receiver cannot be made to do it, then the logstransformprocessor is the next best option as it uses all the libraries that the receivers use. Curious on @djaglowski thoughts on this stance.

So if @linjunzhe goal is better log parsing there may be better options that the transform processor.

But I think the general request of "How can I use the transform processor with some custom functions?" is a valid question we should answer.  Users may need to create their own functions but don't want to bother creating an entirely new component when the transformprocessor would suffice if it knew about the functions.

**Comment by newly12** [comment]
@TylerHelmuth is `logstransformprocessor` intend to replace all `operators` in filelog receiver(aka stanza)? In our use case, we are using regex to match the log body and derive key value pairs when matched, we actually explored the option which is registering customized `operator` in stanza, and leverage its [DSL/expression](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/docs/types/expression.md) for the if/switch part, however we noticed [expr](https://github.com/antonmedv/expr) has lower performance than OTTL(20%~30% decreased), thus we decide to go with OTTL and thus transform processor, if there are better ways of doing this, please do let us know. Thanks.

**Comment by TylerHelmuth** [comment]
@newly12 you just got me all hyped with transformprocessor performance haha

Long term the logstransformprocessor would go away and be replaced by the transform processor, but that is so so so so far off.  In the meantime I don't think @djaglowski has any plan to replace anything in filelog receiver with logstransformprocessor. 

For registering custom components I think it will always require a custom collector because there isn't a good way to add new functions to the transformprocessor that is already built into a collector.  

At the moment I like the idea of being able to expose the transform processor in some way so that custom component creators can say in their new component "use the transform processor, but add these functions" and then add that custom component (which is just the transformprocessor with extra functions) to their custom collector build.

I have not thought yet about how we'd expose the transformprocessor to do that.

**Comment by TylerHelmuth** [comment]
But I think no matter what a user who wants to use custom functions is going to have to make a custom component.  I can't quickly think of a secure way to add custom functions to a transformprocessor thats already built into a collector.

**Comment by kamalmarhubi** [comment]
I'm coming up against this as we start using otelcol more. I want to add some functions that are not really generally applicable enough to add to the `transformprocessor`*.

> But I think no matter what a user who wants to use custom functions is going to have to make a custom component. I can't quickly think of a secure way to add custom functions to a `transformprocessor` thats already built into a collector.

@TylerHelmuth do you mean custom collector or custom component? Given how easy it is to use the [otelcol builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder) to build a custom collector, I think there's a path using something along the lines of the `func Register(name string, function interface{})` API initially suggested.


Here's what I'm imagining:
- I write some custom functions in a package
- that package's `init()` calls the `Register` API
- the otelcol builder is adjusted to allow "extra" packages to be required
  - the way this affects the builder output is that the generated `main.go` adds `_ "example.com/custom/function/package"` imports so that the `Register` functions are called

Then, in the `transformprocessor` config, provide a list of extra functions can be provided by name. These are looked up in the global registry and added the the FactoryMap used to instantiate the various `ottl.Parser`s.

I think this skirts around the "secure way" issue because anyone using it is building their own binary. There's nothing directly distributed by the OpenTelemetry project that has to interface with arbitrary compiled code.

The configuration UX might be a bit messy, but since you can't even use it unless you _very deliberately_ opt in by building a custom collector with functions compiled in, that might be ok?

---
‌* if specific examples are useful, I want to parse a string representing an unsigned 64-bit integer into a SpanID. This doesn't fit in the OTTL data model because pdata has no unsigned integers. I also want to do some parsing of structured header values from logs into attributes. This _might_ be possible with OTTL but it'd be very finicky and hard to test.

**Comment by TylerHelmuth** [comment]
@kamalmarhubi this is an interesting idea.  It would be cool to see a proof of concept presentation at the collector SIG meeting.

**Comment by kamalmarhubi** [comment]
Once my currently-in-progress messy thing is in prod, I'll look at the schedule and see!

**Comment by evan-bradley** [comment]
@frzifus I recall that you were looking into dynamically loading Collector components a while ago. Is there any progress on that, or an issue that we could link to?

**Comment by mx-psi** [comment]
> @frzifus I recall that you were looking into dynamically loading Collector components a while ago. Is there any progress on that, or an issue that we could link to?

I think that would be open-telemetry/opentelemetry-collector/issues/7961

**Comment by frzifus** [comment]
@evan-bradley actually I am planning to pick it up again after my PTO next week. There I would follow the `WASI` path a bit more.

**Comment by newly12** [comment]
I am not sure if I am following this correctly, does the WASI solution mean we will allow the `transformprocessor` to import functions from wsam binary, and it is more preferred because it is a more general and safer solution?

**Comment by kamalmarhubi** [comment]
@TylerHelmuth 
> @kamalmarhubi this is an interesting idea. It would be cool to see a proof of concept presentation at the collector SIG meeting.

I meant to follow up here. I tried to get it to work, but the way generics are used got in the way. I _think_ it might be possible to make it work by forcing external functions to work with their context via a pointer. My understanding of the GC shape stenciling is that all pointer types are "the same" from the generics code generation point of view. I ran out of time for that exploration though.
